Build binaries using Github actions.

remgodow opened the pull request:
A proposal of fix for #46. 
Saves debug binaries for all OS in Github Action Artifact.
Branched from Windows pull request to test Windows OS builds.
Based on [ripgrep](https://github.com/BurntSushi/ripgrep/blob/master/.github/workflows/ci.yml)

imsnif commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,100 @@
+  pull_request:
+  push:
+    branches:
+      - master
```
We use `main` as our main branch

imsnif commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,100 @@
+    strategy:
+      matrix:
+        build:
+          # We test ripgrep on a pinned version of Rust, along with the moving
```
ripgrep => bandwhich :)

imsnif commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,100 @@
+  push:
+    branches:
+      - master
+      - actions
```
I think we can remove this once we're ready to merge and don't need to debug anymore, right?

remgodow replied:
Exactly

imsnif commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,100 @@
+            rust: stable
+          - build: nightly
+            os: ubuntu-18.04
+            rust: nightly
```
Right now on travis, our nightly branch is allowed to fail and we just use it as an FYI. Is there a way to do this here too?

remgodow replied:
Not now unfortunately. I'm not familiar with setting up checks for pull requests, but I guess that we could do ci action with only stable as a gate for pull requests and ci_nightly which would build but would not guard anything.

imsnif reviewed (changes requested):
Awesome. :) I left some minor comments, and after we merge the opensockets PR I think we can definitely go ahead with this one and even use this instead of travis.

remgodow commented:
@imsnif I have added nightly and release actions. You can see the release [here ](https://github.com/remgodow/bandwhich/releases/tag/refs%2Fheads%2Factions). However I noticed that one of the tests fail randomly, could it be a race condition? Here is the stack trace 
```---- tests::cases::ui::pause_by_space stdout ----
thread 'tests::cases::ui::pause_by_space' panicked at 'assertion failed: `(left == right)`
  left: `[Clear, HideCursor, Draw, Flush, Draw, Flush, Draw, Flush, Draw, Flush, Clear, ShowCursor]`,
 right: `[Clear, HideCursor, Draw, Flush, Draw, Flush, Draw, Flush, Clear, ShowCursor]`', src/tests/cases/ui.rs:93:5 ```

imsnif commented:
> @imsnif I have added nightly and release actions. You can see the release [here ](https://github.com/remgodow/bandwhich/releases/tag/refs%2Fheads%2Factions). However I noticed that one of the tests fail randomly, could it be a race condition? Here is the stack trace
> 
> ```
> thread 'tests::cases::ui::pause_by_space' panicked at 'assertion failed: `(left == right)`
>   left: `[Clear, HideCursor, Draw, Flush, Draw, Flush, Draw, Flush, Draw, Flush, Clear, ShowCursor]`,
>  right: `[Clear, HideCursor, Draw, Flush, Draw, Flush, Draw, Flush, Clear, ShowCursor]`', src/tests/cases/ui.rs:93:5 ```
> ```

Yes, I'm afraid we introduced this a few months ago. Mostly on mac, I think. I haven't gotten around to fixing it - I mostly just re-run the tests.

The releases are great! Just need to make sure they match our current release naming and that they are zipped.

imsnif commented:
Hey @remgodow, let me know if there's anything I can pick up on this side. Maybe I can work on the binary release part a bit? (eg. keep our naming convention, make it build/release a generic musl linux file, etc.)?

Also, what do we have left after this? (I know I've asked before somehwere, but I can't seem to find where just now - sorry :) ).
I was also thinking of including some basic textual instructions (or at least a link?) for the winpcap installation, what do you think?

remgodow commented:
@imsnif Sure, feel free to adapt it for project needs. We could add an information for Windows users that WinPcap or NpCap in WinPcap compatible mode is required for bandwhich to work, with a link to NpCap, as WinPcap is depracated. An extra for windows users would be a package for chocolatey/scoop package managers.

remgodow commented:
Scoop looks easy, its uses a simple json with project description and links to binaries as a package manifest. To add it to the repository you make a pull to scoop repo. Here is an example for [ccache](https://github.com/ScoopInstaller/Main/pull/1350/commits/fbdf62cddcc3837623998bd7514f140de0aa99cf)

imsnif commented:
Hey @remgodow, there seems to be a git hitch that I'm not sure I have the permissions to fix here. Could you please merge main into here? I did this locally but am not able to push. I think it's just getting everything from main except the actions themselves.

imsnif commented:
Also, @remgodow - as a side note, I would very much like to add you as an author in Cargo.toml - since you made a very significant contributions. If you'd like that, please add yourself to the bottom of the authors with your email address (if you'd like to include it).

remgodow commented:
The issue with pushing to the branch is caused by release action - right now, it creates /refs/heads/action tags. Therefore it is neccesary to delete release and tag on the fork page to allow pushes. I've added you as collaborator to my fork, I hope that it will allow you to test actions and remove test releases.

imsnif commented:
I played around with this a little. Github actions look cool, but tbh - while I do see the benefit of releasing additional executables, this is too much of a time sink for me right now. Seeing as this isn't blocking anything (to my knowledge) at the moment, I'm going to leave this open and get back to it if the need becomes more pressing. Thanks again for the work on this and the windows port!

remgodow commented:
@imsnif I have adopted release action to current release versioning, and added musl target. Would you need anything else?
Here you can see the effects: https://github.com/remgodow/bandwhich/releases/tag/0.19.7

imsnif commented:
This looks great! Thanks :)

A few things:
1. When unpacking the tar.gz file, I get all the folders of the release in the repository (`target/x86_64-unknown-linux-musl/release`), to fix this I usually tar from the same folder as the original file, but I think there's also a tar flag for this.
2. Would it be possible to make the release a manual step (I think I saw this is possible in github actions) rather than a regex on the commit?
3. How would we get a visual indication for these in the PR? Like we get the pass/fail now with travis and a link to the log? Or will it just happen once we merge?

remgodow commented:
1. Fixed by using working-directory parameter
2. I changed release action so that: it no longer creates release by itself, instead it triggers on "release: created" event - when a release is created using Github, then it builds binaries and uploads them to created release. https://github.com/remgodow/bandwhich/releases/tag/0.20.0
3. When you configure branch protection, it automatically detects github actions and you can select actions which will be required to succeed to allow merging. Checks indicator looks like travis one - https://github.com/remgodow/bandwhich/pull/1 (I guess it does not run because actions are not in main, but you can see the UI).

imsnif commented:
This looks really great. I don't have a lot of time to devote to it right now, but I plan on releasing a new version in the next few days, so will definitely merge this and use it for the release beforehand.

imsnif commented:
So, unfortunately working with github acations is a no-go for me at the moment. I merged this and tried to work with them, but they fail consistently on macos because of some sort of race condition.

I realize that the fault here is in the tests themselves and that a proper fix would be to find the race condition and fix it there, but after spending some time on it I can see that this might not be trivial. At the moment, the travis tests do what we want them to (they sometimes also fail here as well, but much more rarely and almost always run successfully when re-run). So, for our current needs I think they are enough.

Thanks very much for your work on this. I realize you put a lot of effort here. We gave it a good go, but I guess at least in our case, travis is a better option. The actions are still in main but I disabled github actions on this repository so the output will not confuse those wanting to contribute. Maybe in the future I'll give them another go.

remgodow commented:
Sure thing, the benefit of **nigtly and ci** actions over travis was having the artifacts, but its not a deal breaker when it comes to developent builds, hovever there is also a **release** actions which works on all platforms and is triggered manually by creating a release (as it builds and uploads the artifacts, without testing). Therefore I would suggest deleting **nigtly.yaml and ci.yaml**, keeping the **release.yaml**. This way you could benefit from artifacts for releases, while keeping travis for CI/pull requests.

remgodow commented:
By the way, if there is anything I can help with, let me know. I may have some time next week, and I've seen that the windows port caused issues for other systems through the dependencies.

imsnif commented:
> By the way, if there is anything I can help with, let me know. I may have some time next week, and I've seen that the windows port caused issues for other systems through the dependencies.

Uff, yeah - it was a hot mess but I think I took care of it here: https://github.com/imsnif/bandwhich/pull/190 :)

If you'd like something else to work on, this has resurfaced recently: https://github.com/imsnif/bandwhich/issues/150
Also, I've been thinking of adding a flag that would let you manually specify a DNS server instead of autodetecting it.
Only if you feel like it of course. Your contributions are very much welcome.
